### PR TITLE
test(spanner): extend and generalize ConnectionImplTest.QueryOptions

### PR DESCRIPTION
**Pull request by devbww**
Extend `ConnectionImplTest.QueryOptions` to cover all `Client::*` calls
that are affected by `spanner::QueryOptions` and the matching elements of
`spanner::ReadOptions` and `spanner::CommitOptions`, plus the stand-alone
`spanner::RequestPriorityOption`.

Also generalize the protobuf builder so that it will be easier to add
new option fields in the future.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/7141)

### CODE REVIEW

File: google/cloud/spanner/internal/connection_impl_test.cc [hunk 1]
```diff
@@ -83,6 +83,28 @@ using ::testing::Unused;
 
 namespace spanner_proto = ::google::spanner::v1;
 
+// A simple protobuf builder.
```
Comment (coryan) [inline google/cloud/spanner/internal/connection_impl_test.cc:86]: I like this, I may not call it "simple", but I like it.

File: google/cloud/spanner/internal/connection_impl_test.cc [hunk 2]
```diff
@@ -801,23 +867,84 @@ TEST(ConnectionImplTest, QueryOptions) {
...
+              Return(ByMove(absl::make_unique<NiceMock<MockGrpcReader>>())))
+          .RetiresOnSaturation();
+
+      // Commit().
+      EXPECT_CALL(*mock, Commit(_, commit_request_matcher))
+          .Times(1)
+          .RetiresOnSaturation();
+    }
+
+    (void)conn->ExecuteQuery(sql_params);
```
Comment (coryan) [inline google/cloud/spanner/internal/connection_impl_test.cc:939]: Maybe these should be separate tests, but I can see how that would be a lot more boilerplate code. No action needed.

File: google/cloud/spanner/internal/connection_impl_test.cc [hunk 3]
```diff
@@ -706,88 +728,132 @@ TEST(ConnectionImplTest, ExecuteQueryImplicitBeginTransaction) {
...
   struct {
-    // Given this for SqlParams::query_options ...
+    // Given these QueryOptions ...
     spanner::QueryOptions options;
-
-    // ... expect these for ExecuteSqlRequest query_options/request_options.
+    // ... expect these protos.
     spanner_proto::ExecuteSqlRequest::QueryOptions qo_proto;
     spanner_proto::RequestOptions ro_proto;
   } test_cases[] = {
```
Comment (coryan) [inline google/cloud/spanner/internal/connection_impl_test.cc:742]: At some point maybe you would need to split this test cases into separate function, and use `TEST_P()` with the parameter being the values in this array.  Just a thought for the future, right now it is probably six of one, half a dozen of the other.

**Review by coryan — APPROVED** [review]

File: google/cloud/spanner/internal/connection_impl_test.cc (hunk 3 above)
Comment (devbww) [reply google/cloud/spanner/internal/connection_impl_test.cc:742]: Acknowledged.

**Pull request merged**